src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IAsyncEnumerableConverterFactory.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json.Serialization.Converters;
+
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// Converter for streaming <see cref="IAsyncEnumerable{T}" /> values.
+    /// </summary>
+    internal class IAsyncEnumerableConverterFactory : JsonConverterFactory
+    {
+        public override bool CanConvert(Type typeToConvert) => TryGetAsyncEnumerableInterface(typeToConvert, out _);
+
+        public override JsonConverter CreateConverter(Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (!TryGetAsyncEnumerableInterface(typeToConvert, out Type? asyncEnumerableInterface))
+            {
+                Debug.Fail("type not supported by the converter.");
+                throw new Exception();
```

Comment:
Can this ever happen?  Wondering if it can instead just be:
```C#
TryGetAsyncEnumerableInterface(typeToConvert, out Type? asyncEnumerableInterface);
Debug.Assert(asyncEnumerableInterface is not null, $"{typeToConvert} not supported by converter.");
```